[BOL-53] 매치 화면에서 플레이어 선택 화면 구현

prk4224 opened the pull request:
### Reference
- [Jira Ticket](https://yapp22-aos2.atlassian.net/browse/BOL-53)

### 참고 사항

P1: 꼭 반영해주세요 (Request changes)
P2: 적극적으로 고려해주세요 (Request changes)
P3: 웬만하면 반영해 주세요 (Comment)
P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)
P5: 그냥 사소한 의견입니다 (Approve)

prk4224 committed:
Feat: 파일 업로드 로직 구현 (#13)

* refactor: Interceptor  에 ContenetType 제거

* feat: 파일 업로드 Response 생성

* feat: ImageFileApi 생성

* feat: ImageFileApi 의존성 주입

* refactor: DataSourceModule Provide -> Binds

* feat: 파일 업로드 함수 구현 (DataSource)

* feat: 이미지 타입 생성

* feat: 피일 업로드 로직 구현 및 NewGroupUseCase 생성

* feat: NewGroupViewModel 생송

* feat: NewGroupActivity 생성

* feat: 로그인 화면에서 그룹 생성 화면 으로 토큰 값 보내기

* feat: kakaotestViewModel 생성

* feat: 읽기/쓰기 권한 추가

* feat: Glide 빌드 추가

* feat: 갤러리 접근 시 사용되는 데이터 상수화

* feat: getCursor 구현

* feat: setGroupImage 구현

* feat: getRealPathFromURI 구현 (진짜 파일로 만들어주는 함수)

* feat: getResultLauncher 구현

* feat: 권한 가져오기 & 체크 구현

* feat: generateGallery 함수 구현

* feat:checkedGalleryAccess 함수 구현

* feat: 버튼 클릭 시 갤러리 띄우기 & 파일 업로드 함수 호출

prk4224 committed:
refactor:  사용하지 않는 파일 제거

prk4224 committed:
refactor: theme 변경

prk4224 committed:
refactor: 오타 수정

prk4224 committed:
feat: 화면 구현에 필요한 String, color, drawable 추가

prk4224 committed:
feat: 이미지 설정 다이얼 로그 구현

prk4224 committed:
feat: 프로필 설정 다이얼 로그 구현

prk4224 committed:
Feat: Mock Prefix 제거

prk4224 committed:
feat: GroupApi 생성

prk4224 committed:
feat: postCreateGroup 기능 구현

prk4224 committed:
feat: 기능 구현에 필요한 Util 함수 구현

prk4224 committed:
feat: GalleryManager 생성

prk4224 committed:
feat: NewGroup 화면 구현

prk4224 committed:
refactor: safeApiCall 재구성

prk4224 committed:
feat: NewGroupCompleteActivity 구현

prk4224 committed:
refactor: Code Clean Up

prk4224 committed:
refactor: Mock prefix 제거 및 Code Clean Up

prk4224 committed:
refactor: 컨벤션에 맞게 파일 이름 수정

prk4224 committed:
Merge branch 'develop' into feat/BOL-46

prk4224 committed:
BOL-46 refactor: Conflict Solve

prk4224 committed:
feat: navigation 빌드 추가

prk4224 committed:
feat: ImageView setImage Util 함수로 분리

prk4224 committed:
feat: Match Activity  & ViewModel 생성 및 navigation setting

prk4224 committed:
feat: 그룹 생성 완료 버튼 클릭시 MatchActivity 이동 로직 구현

prk4224 committed:
feat:  "게임 결과 기록" 상수 추가

prk4224 committed:
feat: GameItem data class 추가

prk4224 committed:
feat: 게임 선택 화면 구현

prk4224 committed:
feat: 멤버 선태 화면 생성

prk4224 committed:
feat: navigation 에  fragment 추가

prk4224 committed:
refactor: checked 함수 is 로 변경

prk4224 committed:
BOL-54  refactor: 사용하지 않는 data class  제거 및 ktlint

prk4224 committed:
feat: 필요한 drawable, image 추가

prk4224 committed:
refactor: 이름 변경 및 공통 View 이동 (Fragment -> Ativity)

prk4224 committed:
feat: MemberItem 생성

prk4224 committed:
feat: MatchViewModel 데이터 추가

prk4224 committed:
feat: MemberSelectAdapter 생성

prk4224 committed:
feat: MembersAdapter 생성

prk4224 committed:
feat:MemberSelectFragment 구현

prk4224 committed:
feat: 선택된 멤버 X  버튼 클릭시 멤버 리스트에 체크박스 해제 기능 구현

prk4224 committed:
feat: 멤버 검색 & 플레이어 선택 완료 버튼 구현

prk4224 committed:
refactor: Code Clean Up

prk4224 committed:
feat: 멤버 검색 기능 추가

prk4224 committed:
feat: 멤버 스크롤 시 키보드 내려가는 기능 & 코드 클린 업

prk4224 committed:
BOL-53 feat: 검색 버튼 & 취소 버튼 기능 추가

prk4224 committed:
refactor: isEmptyValue -> isInputTextValid 변수 이름 변경

prk4224 committed:
refactoer: 함수 이름 변경 setImage -> loadImage 및 코드 클린 업

prk4224 committed:
refactor: MatchViewModel에 HiltViewModel 추가

prk4224 committed:
refactor: 이름 변경 recyclerAdapter -> gameSelectAdapter

prk4224 committed:
refactor: 코드 클린 업

prk4224 committed:
refactor: 불필요한 라인 제거

prk4224 committed:
fix: player X 버튼 클릭 시 members 체크박스가 제데로 update 되지 않는 버그 수정 및 코드 클린 업

prk4224 committed:
Merge branch 'feat/BOL-54' into feat/BOL-53

# Conflicts:
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/game_select/GameSelectFragment.kt

prk4224 committed:
feat: 게임 리스트 불러오는 기능 구현

prk4224 committed:
fix: getGameList 버그 수정

prk4224 committed:
feat: 네이버 로그인 시 그룹 생성 화면 으로 이동 기능 구현

prk4224 committed:
refactor: Conflict 해경

prk4224 committed:
Merge branch 'develop' into feat/BOL-53

# Conflicts:
#	app/src/main/java/com/yapp/bol/app/di/NetworkModule.kt
#	app/src/main/java/com/yapp/bol/app/di/RepositoryModule.kt
#	buildSrc/src/main/java/com/yapp/bol/Dependencies.kt
#	data/src/main/java/com/yapp/bol/data/datasource/RemoteDataSource.kt
#	data/src/main/java/com/yapp/bol/data/datasource/impl/RemoteDataSourceImpl.kt
#	data/src/main/java/com/yapp/bol/data/mapper/MapperToDomain.kt
#	data/src/main/java/com/yapp/bol/data/remote/GroupApi.kt
#	data/src/main/java/com/yapp/bol/data/repository/RepositoryImpl.kt
#	domain/src/main/java/com/yapp/bol/domain/repository/Repository.kt
#	presentation/src/main/AndroidManifest.xml
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/MatchActivity.kt
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/MatchViewModel.kt
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/game_select/GameSelectAdapter.kt
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/game_select/GameSelectFragment.kt
#	presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt
#	presentation/src/main/java/com/yapp/bol/presentation/viewmodel/login/LoginViewModel.kt
#	presentation/src/main/res/layout/activity_match.xml
#	presentation/src/main/res/layout/fragment_game_selete.xml
#	presentation/src/main/res/layout/fragment_member_select.xml
#	presentation/src/main/res/values/strings.xml

prk4224 committed:
refactor: ktlink 수정

prk4224 committed:
refactor: ktlint 맞추기

prk4224 committed:
refactor: ktlint 수정

prk4224 committed:
refactor: ktllint 수정

eunjjungg commented on domain/src/main/java/com/yapp/bol/domain/usecase/login/MatchUseCase.kt:
```diff
@@ -0,0 +1,15 @@
+package com.yapp.bol.domain.usecase.login
```
p4: package가 login으로 되어있는데 이 부분은 추후에 수정해주시나요?~ 🫠

eunjjungg commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+    companion object {
+        private val diff = object : DiffUtil.ItemCallback<MemberItem>() {
+            override fun areItemsTheSame(oldItem: MemberItem, newItem: MemberItem): Boolean {
+                return oldItem.name == newItem.name
```
p5: 이건 그냥 개인적으로 보다가 아! 그렇구나 한 부분인데 ㅋㅋㅋ 어 왜 id 값으로 비교를 안 하시고 name으로 비교하셨지 생각했습니다..! 근데 생각해보니까 저희 이름이 고유값이라 비교해도 되는군요..!! 🧐👍

prk4224 replied:
악 처음에 MemberItem 만들 때 id 값을 안 넣어서 name 으로 했는데 추후 id로 안 바꿨네요...

eunjjungg commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
+        }
+    }
+
+    private fun setFocusChangeListener(): View.OnFocusChangeListener {
+        return View.OnFocusChangeListener { _, hasFocus ->
+            val image = ContextCompat.getDrawable(
+                requireContext(),
+                if (hasFocus) R.drawable.ic_cancel_gray else R.drawable.ic_search
+            )
+            binding.ivSearchIcon.setImageDrawable(image)
+        }
+    }
```
p5: 저도 이와 같은 기능이 제 부분에 필요해 구현하였는데..! 이렇게 구현할수도 있군요~! 잘 보고 갑니다..ㅎㅎ! 👍

eunjjungg commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
+    }
+
+    private fun showKeyboard(editText: EditText) {
+        if (activity == null && activity?.currentFocus == null) return
```
p4: 궁금한 점인데..! 여기 `activity == null && activity?.currentFocus == null`에서 전자가 null이면 후자는 당연히 null이라고 생각하는데 혹시 제가 잘못 알고 있는 걸까요?! 잘못 안 부분이라면 말씀해주십셔! 🫡

ckrudals replied:
은정님 말씀처럼 `activity == null` 요 조건은 불필요한 것 같네요. `activity?.currentFocus == null` 이 조건만 있어도 충분할 것 같습니다.

prk4224 replied:
그러네요 !  복붙 코드 그대로 사용했네요...하하 수정할게요 😵

eunjjungg reviewed (approved):
스타일 수정에서 구현까지 고생 정말 많으셨습니다~!! ktlint가 참 엄격하네요..하핳... 🥹

ckrudals commented on data/src/main/java/com/yapp/bol/data/model/group/GameDTO.kt:
```diff
@@ -0,0 +1,9 @@
+    val name: String,
+    val minMember: Int,
+    val maxMember: Int,
+    val img: String,
```
p3 : `@serializedname` 을 사용해서 좀더 구체적인 이름을 사용하면 좋을것 같아요 ~

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/NaverTestActivity.kt:
```diff
@@ -59,7 +62,10 @@ class NaverTestActivity : AppCompatActivity() {
     private fun subscribeObservables() {
         viewModel.loginResult.filterNotNull().collectWithLifecycle(this) {
-            showToast("success : $it")
+            if (it.accessToken == Constant.EMPTY_STRING) return@collectWithLifecycle
```
p3 : 전의 리뷰에서도 한번 말했지만 kotlin collection 중 `.isEmpty()` 를 사용하면 개발자가 상수를 따로 지정하지 않아도 원하는 동작을 할 것 같아요 ~

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/MatchViewModel.kt:
```diff
@@ -2,17 +2,91 @@ package com.yapp.bol.presentation.view.match
     private val _toolBarTitle = MutableLiveData(GAME_RESULT_RECORD)
     val toolBarTitle = _toolBarTitle
 
+    private val _members = MutableLiveData<List<MemberItem>>(listOf())
```
p5 : 단순한 궁금증인데 Flow 대신 LiveData 를 사용하시는 이유가 있으실까요?~

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/MatchViewModel.kt:
```diff
@@ -2,17 +2,91 @@ package com.yapp.bol.presentation.view.match
     val toolBarTitle = _toolBarTitle
 
+    private val _members = MutableLiveData<List<MemberItem>>(listOf())
+    val members = _members
```
```suggestion
    private val _members = MutableLiveData<List<MemberItem>>(listOf())
    val members = _members.asLiveData()
```
외부에 보여질 때 immutable 하게 만들기 위해 asLiveData() 로 감싸야해요!
참고 [link](https://velog.io/@changhee09/%EC%95%88%EB%93%9C%EB%A1%9C%EC%9D%B4%EB%93%9C-LiveData#:~:text=%EC%A0%9C%EA%B1%B0%ED%95%A0%20%EC%88%98%20%EC%9E%88%EC%8A%B5%EB%8B%88%EB%8B%A4.-,LiveData%20%EB%A7%8C%EB%93%A4%EA%B8%B0,-LiveData%EB%8A%94%20%EC%9C%84%EC%97%90%EC%84%9C)

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/MatchViewModel.kt:
```diff
@@ -2,17 +2,91 @@ package com.yapp.bol.presentation.view.match
+    companion object {
+        private var allMembers = List(20) {
+            MemberItem(it, "$it. Test", 1)
+        }
```
p3 : 아직 API 연결하기 전 이신거죠 ?

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+    override fun onBindViewHolder(holder: MemberSelectViewHolder, position: Int) {
+        getItem(position)?.let {
+            holder.bind(it)
+        }
```
p3 : `getItem(position)` 이 non-null 인데 let 으로 감쌀 이유가 있어요 ??

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+import com.yapp.bol.presentation.model.MemberItem
+
+class MemberSelectAdapter(
+    private val memberDeleteClickListener: (MemberItem) -> Unit,
```
p3 : `onMemberDelete` 요정도로 만 지어도 외부에 의미 전달이 충분히 될 것 같아요.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
     private var _binding: FragmentMemberSelectBinding? = null
     private val binding get() = checkNotNull(_binding)
```
p3 : `checkNotNull` 이 필요한 검증이에요 ?

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         val gameName = arguments?.getString(GAME_NAME) ?: ""
```
p2 : 전에도 말했지만, viewModel 에서 `savedstatehandle` 를 통해 값을 받아올 수 있어요~ view 에는 코드 가 적으면 적을수록 좋으니 이관하면 좋을 것 같아요!

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
+        if (activity == null && activity?.currentFocus == null) return
+        inputManager.hideSoftInputFromWindow(activity?.currentFocus?.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
+    }
+
```
p3 : keybord 내리는게 다른 화면에서도 공통적으로 사용되어 [utils 에 object](https://github.com/YAPP-Github/22nd-Android-Team-2-Android/pull/48/files#diff-14d9139dcc823d7c7131555bb4eac8a38dde2a4b416f66aa077393704376acb6:~:text=object-,Keyboard,-%7B) 로 만들었습니다 ! 이 pr 이 머지되면 적용하면 좋을 것 같아요 !

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
+        binding.rvMembers.adapter = membersAdapter
+
+        setViewModelObserve()
+        setClickListener()
```
p3 : 함수 추상화 레벨을 통일하면 좋을 것 같아요. 위의 기능들도 함수로 만들면 좋을 것 같아요. 참고 [link](https://javabom.tistory.com/146)

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MemberSelectFragment.kt:
```diff
@@ -17,19 +24,104 @@ class MemberSelectFragment : Fragment() {
+        binding.rvMembers.addOnScrollListener(scrollListener)
+    }
+
+    private fun setViewModelObserve() {
```
p3 : `subscribeObseravles` 정도의 이름이면 될 것 같아요. set 은 외부의 데이터를 내부의 데이터에 설정한다는 느낌인데, viewModel 의 value 를 구독(subscribe) 하고 있으니까요!

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/match/member_select/MembersAdapter.kt:
```diff
@@ -0,0 +1,56 @@
+import com.yapp.bol.presentation.databinding.RvMemberItemBinding
+import com.yapp.bol.presentation.model.MemberItem
+
+class MembersAdapter(
```
p4 : `MemberListAdapter` 정도가 좋을 것 같아요. 멤버 리스트를 `Members` 보단 `MemberList` 을 일반적으로 많이 사용하니 `Members` 는 어색한 느낌이 드네요.

ckrudals commented on presentation/src/main/res/mipmap-hdpi/ic_member_level.png:
p3 : drawable 이 아닌 mipmap 으로 하신 이유가 있나요 ? [일반적으로 다양한 크기 지원](https://developer.android.com/training/multiscreen/screendensities?hl=ko#:~:text=%EC%95%8C%EB%A7%9E%EC%9D%80%20%ED%81%AC%EA%B8%B0%EB%A5%BC%20%EC%84%A0%ED%83%9D%ED%95%A9%EB%8B%88%EB%8B%A4.-,res,-/%0A%20%20%20%20%20%20drawable%2Dxxxhdpi/%0A%20%20%20%20%20%20%20%20awesome)할 때 drawable 을 사용해요!

prk4224 committed:
[ BOL-106] 임시 멤버 추가 기능 & 검색 없을 때 기능 기능 구현 (#44)

* feat: 임시멤버 추가 다이얼 로그 화면 구현

* refactor:  코드 클린 업

* feat: 임시 멤버 추가 다이얼 로그에 이미지 추가

* feat: 임시 멤버 닉네임 중복 체크 API 연결

* faet: 멤버 검색 결과 없을 때 기능 구현

* refactor: Mock Prefix 제거

* refactor: ktlint 수정

* refactor: Query import 추가

* [BOL-106-1] 기능 피드백 수정 (#45)

* refactor: Fragment 애서 사용하는 ViewModel 분리

뒤로가기 등에서 데어트를 지우기 위함.

* feat: 피드백 수정

- 뒤로가기 시 데이터 초기화
- X 버튼 눌렀을 때 키보는 hide 제거
- 멤버 선택시 전체 클릭해도 선택 기능 추가
- 임시멤버 추가 시 키보드 자동 show 기능 추가
- 임시 멤버 이름 입력 뒤 뒤로가기 시 데이터 초기화
- 키보드에 검색 아이콘으로 변경

* refactor: LiveDate Type 추가

* fix: 키도브 Hide 시 dialog가 Show 상태일때문 내리는 로직 수정

* refactpr: multipart Mime Type으로 변경

* refactor: 코드 스타일 수정 및 불필요한 기능 제거

* feat: 디자인 추가

* refactor: 검색결과 없을 때 임시 멤버 클릭 시 키보드 내리기

* refactor: ktlint 수정

* reafactor: ktlint 수정

* [BOL-64] 게임 결과 기록 화면 구현 (#49)

* feat: Image file 생성

* feat: 게임 기록 화면 구현 및 네비게이션 연결

* feat: 게임 결과 기록 Fragment 생성 및 ViewModel 생성

입력 점수에 따른 자동 순위 변경 기능 구현

* refactor: 게임 리스트 불러오는 Url 수정

* refactor: 멤버 검색 결과 없을 때 + 버튼 활성화 상태 체크

* feat: MemberResultItem 생성

* fix: RecyclerView EditText 중복 현상 해결

* feat: 필요한 icon 추가

* feat: 결과 기록 화면 구현

* feat: 기록을 전부 하면 기록 완료 버튼 활성화

* feat: 동점자 처리 기능 추가

* feat: 1위 Text Color 변경

* feat: 필요한 string & 컬러 추가 및 게임 결과 다이얼 로구 구현

* feat: progressbar 구현 및 XML 코드 refactoring

* refacter: Code Refactoring

* feat: Bundle 을 이용해서 값을 넘겨주기 위해 MemberItem 에  Parcelable 상속

* feat:   다이얼 로크 크기 변경하는 로직에서 ResultDialog 일때 고정값 사용하는 로직으로 수정

* retactor: player 순서 변경에서 불필요한 로직 수정

* feat: 사용성 편의를 위한 로직 수정 및 현재 시간

- 키보드 내렸을 때, 다음 버튼 눌렀을 때 순위 자동 수정
- 다음 버튼 클릭 시 스크롤 자동 이동
- 기록 완료 버튼 클릭 시 다이얼로그 띄우기
- 아이템 클릭시 EditText 로 자동 focusing

* refactor: Code Refactoring

* refactor: MOVE_SCROLL_POINT 변경

* ktlint 수정

* refacter: 사용하지 않는 params _ 처리

* refactor: ktlint 수정
